### ISSUE TITLE
make_span(...)?

### ISSUE THREAD

**Issue by rianquinn** [Microsoft/GSL#411]
Is there a reason we don't have a make_span() helper to remove the need for declaring the template type? It would be similar to gsl::finally creating a final_act. It should not be needed once C++17 comes out, but it's going to be a couple of years before C++17 is out and usable by most of the people using this library so having a helper for now would certainly clean up the code a bit. 

Something like (not tested, or compiled)
```cpp
template <class T, class... Args>
inline auto make_span() {
    return span<T>{forward<Args>(args)...};
}
```

There would probably need to be a couple of these to handle the different types of spans, but it should be pretty simple to setup, and then the unit tests could be converted to use the `make_span` version instead to not only simplify them, but ensure the added API is complete

**Comment by MikeGitb** [comment]
What would be the difference compared to the removed  `as_span` function we discussed in  #380?

**Comment by rianquinn** [comment]
huh... I just assumed that as_span was something else, so I guess not much, other than I think `make_span` is more inline with the STL. 

I just added one to our downstream to see how it would work and it seems to work fine (the implementation is a little different than what I put above). It really helps to remove needless verbosity, as I use a lot of type aliases, so my type names can get pretty long.

**Comment by MikeGitb** [comment]
Well - as I wrote in the other issue I also think it would be handy. If I understood neilmacintosh correctly he would probably accept a PR that provides such a function in a separate header.

**Comment by rianquinn** [comment]
I saw that too (I think he said gsl_util), so I can work on upstreaming my changes once I get a chance. I'll likely update the unit tests to use the helpers instead of directly using span, as it tests both at the same time as a result. Also, not all of them will work, for example, the constructors that create a null span won't work since you don't have a type in the function definition to deduce, but for the ones we can do, I will write them, and update the tests as needed.

**Comment by neilmacintosh** [comment]
@rianquinn The suggestion seems pretty reasonable to me. There is certainly plenty of precedent in the standard library for `make_` names and I do understand the convenience of the syntax. However, as I think I observed in a previous issue, the fancy new template constructor deduction feature (not the technical term) in C++17 may reduce the syntactic need for `make_span`.

Nonetheless, I think the addition would be a good one to help people out while we wait for C++17 to be finalized and then propagate through the world.

**Comment by MikeGitb** [comment]
Stupid question: has anyone investigated, whether that c++17 feature would actually work for gsl::span's constructors, as they are overloaded, templated and use somewhat complicated enable_if expressions?

**Comment by neilmacintosh** [comment]
@MikeGitb there's no such thing as a stupid question, unless it's unasked ;-)

I had a quick look and I believe that it should work for many cases. The feature also allows for explicit deduction guides, which would help a lot. However, I wouldn't guarantee that it will always work everywhere (we already have some ambiguities)...or that everyone wants to use the syntax associated with that feature.

That's why I'm happy to look at `make_span` now: I can imagine there'll be quite a few places people will want that alternative.

**Comment by viboes** [comment]
I believe that all the overloads proposed for make_span would be possible with deduction guides.

BTW the feature is called "template argument deduction for class templates (constructors)".

**Comment by neilmacintosh** [comment]
Thanks @viboes, I thought I was just being super-forgetful, but that really is a lot of letters. My chances of remembering it in the future are pretty close to zero, but you have done me a great service...now I can return here and copy-paste it from your comment! ;-)

**Comment by neilmacintosh** [comment]
This has now been addressed by PR #413. Thanks for the PR @rianquinn!
